maps.Rmd

```diff
@@ -299,52 +299,140 @@ ggplot(dawson[-69]) +
 
-mi_cities <- tbl_df(maps::us.cities) %>%
-  filter(country.etc == "MI") %>%
-  select(-country.etc, lon = long) 
+To plot the `sat_vis` data in ggplot2, we can use the `geom_stars()` function provided by the stars package. A minimal plot might look like this:
 
-ggmap::ggmap(mi_raster) +
-  geom_point(aes(size = pop), mi_cities, colour = "red") +
-  scale_size_area()
+`r columns(2, 1/2, max_width = 2)`
+```{r}
+ggplot() + 
+  geom_stars(data = sat_vis) + 
+  coord_equal()
 ```
 
-The nice thing about the `ggmap()` function is that the output is a ggplot object, so adding metadata to the plot is accomplished in the same way one would add layers to any other plot. This is illustrated in the right side plot above.
+The first thing to note is that `geom_stars()` requires the `data` argument to be a stars object, and its job is to take the stars object and render it using `geom_raster()` (the default) or `geom_sf()`. The choice of geom can be controlled by the user: you can set `sf = TRUE` to use `geom_sf()`, but note that rendering will be slow. he only way that `geom_stars()` is able to call `geom_sf()` this way is by converting the pixel map to a collection of polygons that `geom_sf()` understands, which is a much slower process than using `geom_raster()`. 
 
-<!-- If you have raster data from the raster package, you can convert it to the form needed by ggplot2 with the following code: -->
+The second point to make about this image is that the raster data is mapped to the fill aesthetic, and the blue shading in the satellite image above is determined by the ggplot2 scale, not the image itself. That is, although `sat_vis` contains three bands, the plot above only displays the first one, and the raw data values (which range from 0 to 255) are mapped onto the default blue palette that ggplot2 uses for continuous data. To see what the image file "really" looks like we can separate the bands using `facet_wrap()`:
 
-<!-- ```{r, eval = FALSE} -->
-<!-- df <- as.data.frame(raster::rasterToPoints(x)) -->
-<!-- names(df) <- c("lon", "lat", "x") -->
+```{r}
+ggplot() + 
+  geom_stars(data = sat_vis, show.legend = FALSE) +
+  facet_wrap(vars(band)) + 
+  coord_equal() + 
+  scale_fill_gradient(low = "black", high = "white")
+```
 
-<!-- ggplot(df, aes(lon, lat)) + -->
-<!--   geom_raster(aes(fill = x)) -->
-<!-- ``` -->
+One limitation to displaying only the raw image is that it is not easy to work out where the relevant landmasses are, and we may wish to overlay the satellite data with the `oz_states` vector map to show the outlines of Australian political entities. However, some care is required in doing so because the two data sources are associated with different coordinate reference systems, which we could verify by comparing `st_crs(sat_vis)` to `st_crs(oz_states)`.
```

Comment:
Hmmm. Why doesn't this happen automatically?

Reply:
I'm only guessing, but I *think* it's this:

- `sat_vis` has scalar class `stars` so I `ggplot()` throws an error if I use it as the global data, so I'm passing it as local data to the `geom_stars()` layer
- the call to `geom_stars()` is passed to `geom_raster()` so my guess is that by the time the plot method is called this layer doesn't have a CRS?
- so then `oz_states` is the first layer that specifies a CRS that the plot method knows about 
- but the CRS for `oz_states` is all wrong for the raster layer, so if I don't manually call `st_transform()` the vector layer shows up as a white dot in the middle of the image

Assuming I'm not completely misunderstanding, I'm sort of tempted to sweep this one under the carpet?

Reply:
Yeah, I'd be tempted to sweep this under the rug